amoro-ams/src/main/java/org/apache/amoro/server/table/DefaultTableRuntime.java

```diff
@@ -0,0 +1,162 @@
+import org.apache.amoro.shade.zookeeper3.org.apache.curator.shaded.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+public class DefaultTableRuntime extends StatedPersistentBase implements SupportsProcessPlugins {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultTableRuntime.class);
+
+  private final ServerTableIdentifier tableIdentifier;
+  private final DefaultOptimizingState optimizingState;
+  private final Map<Action, TableProcessContainer> processContainerMap = Maps.newConcurrentMap();
+
+  public DefaultTableRuntime(
+      ServerTableIdentifier tableIdentifier,
+      TableRuntimeHandler tableHandler,
+      Map<String, String> properties) {
+    Preconditions.checkNotNull(tableIdentifier, "ServerTableIdentifier must not be null.");
+    Preconditions.checkNotNull(tableHandler, "TableRuntimeHandler must not be null.");
+    this.tableIdentifier = tableIdentifier;
+    this.optimizingState = new DefaultOptimizingState(this, tableHandler, properties);
+  }
+
+  public DefaultTableRuntime(TableRuntimeMeta tableRuntimeMeta, TableRuntimeHandler tableHandler) {
+    Preconditions.checkNotNull(tableRuntimeMeta, "TableRuntimeMeta must not be null.");
+    Preconditions.checkNotNull(tableHandler, "TableRuntimeHandler must not be null.");
+
+    this.tableIdentifier =
+        ServerTableIdentifier.of(
+            tableRuntimeMeta.getTableId(),
+            tableRuntimeMeta.getCatalogName(),
+            tableRuntimeMeta.getDbName(),
+            tableRuntimeMeta.getTableName(),
+            tableRuntimeMeta.getFormat());
+    this.optimizingState = new DefaultOptimizingState(this, tableRuntimeMeta, tableHandler);
+  }
+
+  public DefaultOptimizingState getOptimizingState() {
+    return optimizingState;
+  }
+
+  public void recover(OptimizingProcess optimizingProcess) {
+    optimizingState.recover(optimizingProcess);
+  }
+
+  public void registerMetric(MetricRegistry metricRegistry) {
+    optimizingState.registerMetric(metricRegistry);
+  }
+
+  public void dispose() {
+    optimizingState.dispose();
+  }
+
+  @Override
+  public AmoroProcess<? extends TableProcessState> trigger(Action action) {
+    return Optional.ofNullable(processContainerMap.get(action))
+        .map(container -> container.trigger(action))
+        // Define a related exception
+        .orElseThrow(() -> new IllegalArgumentException("No ProcessFactory for action " + action));
+  }
+
+  @Override
+  public void install(Action action, ProcessFactory<? extends TableProcessState> processFactory) {
+    if (processContainerMap.putIfAbsent(action, new TableProcessContainer(processFactory))
+        == null) {
+      throw new IllegalStateException("ProcessFactory for action " + action + " already exists");
```

Comment:
I think this is inverted. Should be `!= null` instead.
```suggestion
    if (processContainerMap.putIfAbsent(action, new TableProcessContainer(processFactory))
        != null) {
      throw new IllegalStateException("ProcessFactory for action " + action + " already exists");
```